src/theory/strings/output_channel.cpp

```diff
@@ -0,0 +1,448 @@
+  d_pending.push_back(eq);
+  d_pending_exp[eq] = eq_exp;
+  d_keep.insert(eq);
+  d_keep.insert(eq_exp);
+}
+
+bool OutputChannelStrings::sendSplit(Node a, Node b, const char* c, bool preq)
+{
+  Node eq = a.eqNode(b);
+  eq = Rewriter::rewrite(eq);
+  if (eq.isConst())
+  {
+    return false;
+  }
+  NodeManager* nm = NodeManager::currentNM();
+  Node lemma_or = nm->mkNode(OR, eq, nm->mkNode(NOT, eq));
+  Trace("strings-lemma") << "Strings::Lemma " << c << " SPLIT : " << lemma_or
+                         << std::endl;
+  d_pending_lem.push_back(lemma_or);
+  sendPhaseRequirement(eq, preq);
+  return true;
+}
+
+void OutputChannelStrings::sendPhaseRequirement(Node lit, bool pol)
+{
+  d_pending_req_phase[lit] = pol;
+}
+
+Node OutputChannelStrings::getRepresentative(Node t)
+{
+  if (d_ee.hasTerm(t))
+  {
+    return d_ee.getRepresentative(t);
+  }
+  return t;
+}
+
+bool OutputChannelStrings::hasTerm(Node a) { return d_ee.hasTerm(a); }
+
+bool OutputChannelStrings::areEqual(Node a, Node b)
+{
+  if (a == b)
+  {
+    return true;
+  }
+  else if (hasTerm(a) && hasTerm(b))
+  {
+    return d_ee.areEqual(a, b);
+  }
+  return false;
+}
+
+bool OutputChannelStrings::areDisequal(Node a, Node b)
+{
+  if (a == b)
+  {
+    return false;
+  }
+  if (hasTerm(a) && hasTerm(b))
+  {
+    Node ar = d_ee.getRepresentative(a);
+    Node br = d_ee.getRepresentative(b);
+    return (ar != br && ar.isConst() && br.isConst())
+           || d_ee.areDisequal(ar, br, false);
+  }
+  Node ar = getRepresentative(a);
+  Node br = getRepresentative(b);
+  return ar != br && ar.isConst() && br.isConst();
+}
+
+void OutputChannelStrings::doPendingFacts()
+{
+  size_t i = 0;
+  while (!hasConflict() && i < d_pending.size())
+  {
+    Node fact = d_pending[i];
+    Node exp = d_pending_exp[fact];
+    if (fact.getKind() == AND)
+    {
+      for (const Node& lit : fact)
+      {
+        bool polarity = lit.getKind() != NOT;
+        TNode atom = polarity ? lit : lit[0];
+        d_parent.assertPendingFact(atom, polarity, exp);
+      }
+    }
+    else
+    {
+      bool polarity = fact.getKind() != NOT;
+      TNode atom = polarity ? fact : fact[0];
+      d_parent.assertPendingFact(atom, polarity, exp);
+    }
+    i++;
+  }
+  d_pending.clear();
+  d_pending_exp.clear();
+}
+
+void OutputChannelStrings::doPendingLemmas()
+{
+  if (!hasConflict())
+  {
+    for (const Node& lc : d_pending_lem)
+    {
+      Trace("strings-pending") << "Process pending lemma : " << lc << std::endl;
+      d_out.lemma(lc);
+    }
+    for (const std::pair<const Node, bool>& prp : d_pending_req_phase)
+    {
+      Trace("strings-pending") << "Require phase : " << prp.first
+                               << ", polarity = " << prp.second << std::endl;
+      d_out.requirePhase(prp.first, prp.second);
+    }
+  }
+  d_pending_lem.clear();
+  d_pending_req_phase.clear();
+}
+
+bool OutputChannelStrings::hasConflict() const { return d_parent.d_conflict; }
```

Comment:
Wouldn't it be more natural to move `d_parent.d_conflict` to `OutputChannelStrings`?

Reply:
There would be a circular include.

Reply:
Hm, I am not quite following. `theory_strings.h` already includes `inference_manager.h` https://github.com/CVC4/CVC4/pull/3076/files#diff-452ff9fd1723599f909e96f23d6f366dR27, no?

Reply:
Yes, so inference_manager cannot include theory_strings.h. Hence, this function needs to be in the cpp of inference_manager.

Reply:
I think there was a bit of a misunderstanding here :) I was suggesting to move the `d_conflict` variable to `InferenceManager` ;) Not too important, we can potentially address this in a later PR.

Reply:
Ah.  Yes I considered this.

I think in the end I'm going to split another class off of TheoryStrings, which relates to the "state" of the solver, which includes things like read queries to equality engine and the normal form information.  Otherwise I think the inference manager will become too big of a class.  The intention of this class is to mostly act as a means of performing the "output" of strategies and not means of tracking the state of the solver.

I am following a very similar approach in refactoring sets, see InferenceManager, SetsState (https://github.com/CVC4/CVC4/compare/master...ajreynol:setsCardSplit).